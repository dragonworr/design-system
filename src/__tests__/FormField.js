import React from 'react'
import renderer from 'react-test-renderer'
import { renderIntoDocument } from 'react-dom/test-utils'
import 'jest-styled-components'
import { FormField, Icon, Input, Label, Select } from '..'

describe('FormField', () => {
  test('renders', () => {
    const json = renderer
      .create(
        <FormField>
          <Label htmlFor="email">Email Address</Label>
          <Input id="email" />
        </FormField>
      )
      .toJSON()
    expect(json).toMatchSnapshot()
  })

  test('renders with Icon', () => {
    const json = renderer
      .create(
        <FormField>
          <Label htmlFor="email">Email Address</Label>
          <Icon id="email" />
          <Input id="email" />
        </FormField>
      )
      .toJSON()
    expect(json).toMatchSnapshot()
  })

  test('renders with Select ', () => {
    const json = renderer
      .create(
        <FormField>
          <Select />
        </FormField>
      )
      .toJSON()
    expect(json).toMatchSnapshot()
  })

  test('renders with Select and Icon', () => {
    const json = renderer
      .create(
        <FormField>
          <Icon name="email" />
          <Select />
        </FormField>
      )
      .toJSON()
    expect(json).toMatchSnapshot()
  })

  test('renders with showLabel prop', () => {
    const json = renderer
      .create(
        <FormField showLabel>
          <Label htmlFor="email">Email</Label>
          <Icon name="email" />
          <Input id="email" />
        </FormField>
      )
      .toJSON()
    expect(json).toMatchSnapshot()
  })

  test('shows label with field value', () => {
    const json = renderer
      .create(
        <FormField>
          <Label htmlFor="email">Email</Label>
          <Icon name="email" />
          <Input id="email" value="hello@example.com" />
        </FormField>
      )
      .toJSON()
    const [label] = json.children
    expect(label.props.style.opacity).toBe(1)
  })

  /*
  test('it correctly places an aria-label with the placeholder value before user has interacted with the input', () => {
    const test = mount(
      <FormField>
        <Label>A Label</Label>
        <Icon name="email" />
        <Input id="with-both-icons" placeholder="placeholder text" />
        <Icon name="email" />
      </FormField>
    )

    // before user interaction the Label element should not be rendered
    let label = test.find(Label)
    expect(label.length).toBe(0)

    let input = test.find(Input)
    expect(input.getDOMNode().getAttribute('aria-label')).toBe(
      'placeholder text'
    )
  })

  */

  test('warns when there is no form field', () => {
    const spy = jest.spyOn(global.console, 'error')
    renderIntoDocument(
      <FormField>
        <Label>Email Address</Label>
      </FormField>
    )
    expect(spy).toHaveBeenCalledTimes(1)
    spy.mockRestore()
  })

  test('warns when there is no label', () => {
    const spy = jest.spyOn(global.console, 'error')
    renderIntoDocument(
      <FormField>
        <Input />
      </FormField>
    )
    expect(spy).toHaveBeenCalledTimes(1)
    spy.mockRestore()
  })

  /*
  test('it triggers a prop-type warning when any element besides Label, Icon, and Input is provided as a child.', () => {
    console.error = jest.fn()

    const test = mount(
      <FormField onChange={() => {}}>
        <span>Extra stuff not supported</span>
        <Label>A Label</Label>

        <Icon name="email" />
        <Icon name="email" />
        <Input id="with-both-icons" />
        <Icon name="email" />
      </FormField>
    )

    expect(
      console.error.mock.calls
        .toString()
        .indexOf('is not a valid child for') !== -1
    )
    console.error.mockRestore()
  })
  test('it triggers a prop-type warning when more than one Label is provided', () => {
    console.error = jest.fn()

    const test = mount(
      <FormField onChange={() => {}}>
        <Label>A Label</Label>
        <Label>Another Label</Label>
        <Input id="with-both-icons" />
      </FormField>
    )

    expect(
      console.error.mock.calls
        .toString()
        .indexOf("Exactly 0 or 1 'Label' children should be supplied to ") !==
        -1
    )
    console.error.mockRestore()
  })
  test('it triggers a prop-type warning when there is more than one icon before or after the Input element', () => {
    console.error = jest.fn()

    const test = mount(
      <FormField onChange={() => {}}>
        <span>Extra stuff not supported</span>
        <Label>A Label</Label>

        <Icon name="email" />
        <Icon name="email" />
        <Input id="with-both-icons" />
      </FormField>
    )

    expect(
      console.error.mock.calls
        .toString()
        .indexOf('component must be positioned between them as children of') !==
        -1
    )
    console.error.mockRestore()
  })
  */
})
