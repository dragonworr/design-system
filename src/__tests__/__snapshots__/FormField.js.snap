// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormField renders 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c2 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  display: block;
  width: 100%;
  font-family: inherit;
  font-size: 14px;
  color: inherit;
  background-color: transparent;
  border-radius: 2px;
  border-width: 0px;
  border-style: solid;
  border-color: #d1d6db;
  padding-top: 14px;
  padding-bottom: 14px;
  padding-left: 12px;
  padding-right: 12px;
  margin: 0;
  border-color: #d1d6db;
  box-shadow: 0 0 0 1px #d1d6db;
}

.c2::-webkit-input-placeholder {
  color: #687B8E;
}

.c2::-moz-placeholder {
  color: #687B8E;
}

.c2:-ms-input-placeholder {
  color: #687B8E;
}

.c2::placeholder {
  color: #687B8E;
}

.c2::-ms-clear {
  display: none;
}

.c2:focus {
  outline: 0;
  border-color: #007aff;
  box-shadow: 0 0 0 2px #007aff;
}

.c0 {
  font-size: 10px;
  -webkit-letter-spacing: 0.2px;
  -moz-letter-spacing: 0.2px;
  -ms-letter-spacing: 0.2px;
  letter-spacing: 0.2px;
  display: block;
  width: 100%;
  margin: 0;
  margin-bottom: -20px;
  margin-left: 12px;
  padding-top: 6px;
  font-size: 10px;
  color: #687B8E;
  font-weight: 700;
}

<div>
  <label
    className="c0"
    color="gray"
    fontSize={10}
    fontWeight="bold"
    htmlFor="email"
    style={
      Object {
        "height": 20,
        "opacity": 0,
        "transitionDuration": ".1s",
        "transitionProperty": "opacity",
      }
    }
  >
    Email Address
  </label>
  <div
    className="c1"
  >
    <input
      className="c2"
      id="email"
      style={
        Object {
          "paddingBottom": 14,
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "paddingTop": 14,
          "transitionDuration": ".1s",
          "transitionProperty": "padding-top, padding-bottom",
        }
      }
    />
  </div>
</div>
`;

exports[`FormField renders with Icon 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c2 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  display: block;
  width: 100%;
  font-family: inherit;
  font-size: 14px;
  color: inherit;
  background-color: transparent;
  border-radius: 2px;
  border-width: 0px;
  border-style: solid;
  border-color: #d1d6db;
  padding-top: 14px;
  padding-bottom: 14px;
  padding-left: 12px;
  padding-right: 12px;
  margin: 0;
  border-color: #d1d6db;
  box-shadow: 0 0 0 1px #d1d6db;
}

.c2::-webkit-input-placeholder {
  color: #687B8E;
}

.c2::-moz-placeholder {
  color: #687B8E;
}

.c2:-ms-input-placeholder {
  color: #687B8E;
}

.c2::placeholder {
  color: #687B8E;
}

.c2::-ms-clear {
  display: none;
}

.c2:focus {
  outline: 0;
  border-color: #007aff;
  box-shadow: 0 0 0 2px #007aff;
}

.c0 {
  font-size: 10px;
  -webkit-letter-spacing: 0.2px;
  -moz-letter-spacing: 0.2px;
  -ms-letter-spacing: 0.2px;
  letter-spacing: 0.2px;
  display: block;
  width: 100%;
  margin: 0;
  margin-bottom: -20px;
  margin-left: 40px;
  padding-top: 6px;
  font-size: 10px;
  color: #687B8E;
  font-weight: 700;
}

<div>
  <label
    className="c0"
    color="gray"
    fontSize={10}
    fontWeight="bold"
    htmlFor="email"
    style={
      Object {
        "height": 20,
        "opacity": 0,
        "transitionDuration": ".1s",
        "transitionProperty": "opacity",
      }
    }
  >
    Email Address
  </label>
  <div
    className="c1"
  >
    <input
      className="c2"
      id="email"
      style={
        Object {
          "paddingBottom": 14,
          "paddingLeft": 40,
          "paddingRight": undefined,
          "paddingTop": 14,
          "transitionDuration": ".1s",
          "transitionProperty": "padding-top, padding-bottom",
        }
      }
    />
  </div>
</div>
`;

exports[`FormField renders with Select  1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c3 {
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  margin-left: -32px;
  height: 24px;
  width: 24px;
  color: #687B8E;
  pointer-events: none;
}

.c2 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  display: block;
  width: 100%;
  font-family: inherit;
  color: inherit;
  background-color: transparent;
  border-radius: 2px;
  border-width: 1px;
  border-style: solid;
  border-color: #d1d6db;
  margin: 0px;
  padding-left: 12px;
  padding-right: 32px;
  padding-top: 14px;
  padding-bottom: 14px;
  font-size: 14px;
}

.c2:focus {
  outline: none;
  border-color: #007aff;
  box-shadow: 0 0 0 1px #007aff;
}

<div>
  <div
    className="c0"
  >
    <div
      className="c1"
      width={1}
    >
      <select
        className="c2"
        fontSize={1}
        style={
          Object {
            "paddingBottom": 14,
            "paddingLeft": undefined,
            "paddingRight": undefined,
            "paddingTop": 14,
            "transitionDuration": ".1s",
            "transitionProperty": "padding-top, padding-bottom",
          }
        }
      />
      <svg
        className="c3"
        fill="currentcolor"
        viewBox="0 0 24 24"
      >
        <path
          d="M7.4,8l4.6,4.6L16.6,8L18,9.4l-6,6l-6-6L7.4,8z"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`FormField renders with Select and Icon 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  height: 24px;
  width: 24px;
}

.c4 {
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  margin-left: -32px;
  height: 24px;
  width: 24px;
  color: #687B8E;
  pointer-events: none;
}

.c3 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  display: block;
  width: 100%;
  font-family: inherit;
  color: inherit;
  background-color: transparent;
  border-radius: 2px;
  border-width: 1px;
  border-style: solid;
  border-color: #d1d6db;
  margin: 0px;
  padding-left: 12px;
  padding-right: 32px;
  padding-top: 14px;
  padding-bottom: 14px;
  font-size: 14px;
}

.c3:focus {
  outline: none;
  border-color: #007aff;
  box-shadow: 0 0 0 1px #007aff;
}

<div>
  <div
    className="c0"
  >
    <svg
      className="c1"
      fill="currentcolor"
      style={
        Object {
          "alignSelf": "center",
          "flex": "none",
          "marginLeft": 8,
          "marginRight": -32,
          "pointerEvents": "none",
        }
      }
      viewBox="0 0 24 24"
    >
      <path
        d="M20,4H4C2.9,4,2,4.9,2,6v12c0,1.1,0.9,2,2,2h16c1.1,0,2-0.9,2-2V6C22,4.9,21.1,4,20,4z M20,8
      	l-8,5L4,8V6l8,5l8-5V8z"
      />
    </svg>
    <div
      className="c2"
      width={1}
    >
      <select
        className="c3"
        fontSize={1}
        style={
          Object {
            "paddingBottom": 14,
            "paddingLeft": 40,
            "paddingRight": undefined,
            "paddingTop": 14,
            "transitionDuration": ".1s",
            "transitionProperty": "padding-top, padding-bottom",
          }
        }
      />
      <svg
        className="c4"
        fill="currentcolor"
        viewBox="0 0 24 24"
      >
        <path
          d="M7.4,8l4.6,4.6L16.6,8L18,9.4l-6,6l-6-6L7.4,8z"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`FormField renders with showLabel prop 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c2 {
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  height: 24px;
  width: 24px;
}

.c3 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  display: block;
  width: 100%;
  font-family: inherit;
  font-size: 14px;
  color: inherit;
  background-color: transparent;
  border-radius: 2px;
  border-width: 0px;
  border-style: solid;
  border-color: #d1d6db;
  padding-top: 14px;
  padding-bottom: 14px;
  padding-left: 12px;
  padding-right: 12px;
  margin: 0;
  border-color: #d1d6db;
  box-shadow: 0 0 0 1px #d1d6db;
}

.c3::-webkit-input-placeholder {
  color: #687B8E;
}

.c3::-moz-placeholder {
  color: #687B8E;
}

.c3:-ms-input-placeholder {
  color: #687B8E;
}

.c3::placeholder {
  color: #687B8E;
}

.c3::-ms-clear {
  display: none;
}

.c3:focus {
  outline: 0;
  border-color: #007aff;
  box-shadow: 0 0 0 2px #007aff;
}

.c0 {
  font-size: 10px;
  -webkit-letter-spacing: 0.2px;
  -moz-letter-spacing: 0.2px;
  -ms-letter-spacing: 0.2px;
  letter-spacing: 0.2px;
  display: block;
  width: 100%;
  margin: 0;
  margin-bottom: -20px;
  margin-left: 40px;
  padding-top: 6px;
  font-size: 10px;
  color: #687B8E;
  font-weight: 700;
}

<div>
  <label
    className="c0"
    color="gray"
    fontSize={10}
    fontWeight="bold"
    htmlFor="email"
    style={
      Object {
        "height": 20,
        "opacity": 1,
        "transitionDuration": ".1s",
        "transitionProperty": "opacity",
      }
    }
  >
    Email
  </label>
  <div
    className="c1"
  >
    <svg
      className="c2"
      fill="currentcolor"
      style={
        Object {
          "alignSelf": "center",
          "flex": "none",
          "marginLeft": 8,
          "marginRight": -32,
          "pointerEvents": "none",
        }
      }
      viewBox="0 0 24 24"
    >
      <path
        d="M20,4H4C2.9,4,2,4.9,2,6v12c0,1.1,0.9,2,2,2h16c1.1,0,2-0.9,2-2V6C22,4.9,21.1,4,20,4z M20,8
      	l-8,5L4,8V6l8,5l8-5V8z"
      />
    </svg>
    <input
      className="c3"
      id="email"
      style={
        Object {
          "paddingBottom": 8,
          "paddingLeft": 40,
          "paddingRight": undefined,
          "paddingTop": 20,
          "transitionDuration": ".1s",
          "transitionProperty": "padding-top, padding-bottom",
        }
      }
    />
  </div>
</div>
`;
